Automattic/woocommerce-payments
# feat: tokenized cart PRB PDP session handler

Fixes https://github.com/Automattic/woocommerce-payments/issues/8841

#### Changes proposed in this Pull Request









Adding capability to pay for subscriptions & associating orders placed by the current logged-in customer.


https://github.com/user-attachments/assets/a8ab4c47-f87d-4c1c-9aa7-805f88d07493

When an order is placed with the PRB on a product page, the customer's main cart should remain unaltered.
So, if the customer placed an order for one product (but had other items in their cart), they should be able to check out only the single product.

After discussing with the Blocks team, the solution we concluded with was to implement a custom SessionHandler.

#### Testing instructions





- Disable ECE with the following snippet:
```php
add_filter( 'pre_option__wcpay_feature_stripe_ece', function () {
    return '0'; 
}, 100 );
```
- Enable the "Enable Cart-Token implementation for PRBs" flag in your dev tools (you'll need to update them)
- Ensure you have Google Pay/Apple Pay enabled in the merchant's settings
- Ensure you have WC Subscriptions installed
- As a logged-in customer, add some random products to the cart (we'll need this to later validate that the main customer cart isn't affected)
- Go to a subscription's product page
- Click on the GooglePay/ApplePay button to purchase the subscription
- A few things can happen:
  - If you already placed a subscription order on this site and you're logged in, the order should be placed (and the backend will associate the order to your account)
  - If you already placed a subscription order on this site and you're _not_ logged in, the order will fail with an error message in the back of the page (the behavior is a bit janky - but it's the same behavior as the non-tokenized cart implementation)
<img width="894" alt="Screenshot 2024-07-18 at 2 40 30 PM" src="https://github.com/user-attachments/assets/79e44bf4-4ea4-4c7a-8a80-6b329210f5d1">
  - If you haven't placed a subscription order on this site, the new subscription will be created and the backend will associate the order to your account (if you have one)


-------------------


**Post merge**
